### ISSUE TITLE
dashing (docset generation) is disabled due to taking too long

### ISSUE THREAD

**Issue by christopherfujino** [flutter/flutter#60646]
I noticed this started happening 6/29/20. Here's [an example](https://cirrus-ci.com/task/6625596312649728) (on that PR it timed out twice in a row).

**Comment by jonahwilliams** [comment]
the last few times this happened, it was an issue on the firebase side of things (I'm not sure if we're still using it). FYI @gspencergoog or @kf6gpe might still have some contacts here

**Comment by christopherfujino** [comment]
Perhaps related, it would also be nice to have more verbose logging, rather than just:

```
Zipping Flutter offline docs archive.
Building Flutter docset.
Timed out!
```

**Comment by christopherfujino** [comment]
I'll make that a separate issue: https://github.com/flutter/flutter/issues/60647

**Comment by christopherfujino** [comment]
> the last few times this happened, it was an issue on the firebase side of things (I'm not sure if we're still using it). FYI @gspencergoog or @kf6gpe might still have some contacts here

So, due to the output, it looks like the problem is dashing building the docset, before the firebase upload.

**Comment by zanderso** [comment]
Raising this to P0 since this has started timing out consistently. Assigning tvolkert@ to reflect that he is currently taking a stab at this.

**Comment by tvolkert** [comment]
With https://github.com/flutter/flutter/pull/62251 merged, we have a [docs shard run with full logging on the `dashing` command](https://api.cirrus-ci.com/v1/task/6387415059267584/logs/main.log).  The results:

| Job           | Time taken    | Completed | Logs Produced (size) |
| ------------- | ------------- | --------- | -------------------- |
| Dartdoc       | 14:35         | Y         | 253.30 KB            |
| Offline ZIP   | 12:09         | Y         |  68.00 B             |
| Dashing       | 33:16         | N (timed out) |  20.35 MB            |
| Firebase Deploy | n/a | N (not started) | n/a |

There's nothing in the dashing logs that looks problematic - it seems to simply be taking too long.

__Ideas__:

1. Try to find a way to speed up the `dashing` process.  @imWildCat given your help with the original wiring up of this, any chance you have insight here?
2. Try to separate the docset generation into its own shard in Cirrus, and probably separate out the firebase deployment into its own shard while we were at it.  We'd need to make sure those shards didn't run until doc generation was complete, and that they ran in the same container so that the docs were still there.  I haven't yet looked into whether Cirrus supports this.  @fkorotkov can you advise?
3. See if we can raise the Cirrus timeout last 60 minutes for the docs shard.
4. Disable docset generation for now until we can migrate this task from Cirrus to LUCI.  I'd rather avoid this option since developers use the docsets we publish, and I wouldn't want to go any real length of time without them being published.
5. Something I haven't considered?

**Comment by tvolkert** [comment]
Per discussion on Discord (falls under option 1)

@zanderso:
> Looks like it was making progress when it timed out. Comparing the CPU and Memory charts between that run and the last successful run, looks like it becomes IO bound and hits a wall. I wonder if it's hitting a real disk or a memfs. Guessing a real disk and  maybe sharing with another disk bound task.

@fkorotkov 
> Cirrus uses a normal persistent disk on GCP and IO can be shared with other tasks on the node. I can expose an option to use tmpfs if you have high IO tasks.
There is already such option for private GKE tasks. :thinking:

**Comment by fkorotkov** [comment]
@tvolkert coukd you please clarify option 2 a bit? You want to generate docs in separate task, wait for all other tasks to finish, do a deployment from the generated docs in the first container?

**Comment by tvolkert** [comment]
Yes, the idea would be to have separate tasks for the following:

1. doc generation
1. doc zipping for offline docs
1. docset generation (dashing)
1. firebase deployment of all of the above

I'd want to structure the tasks such that 2 & 3 would depend on 1, and 4 would depend on 2 & 3 -- and I'd want to make sure that any task would be able to see the files that were written by tasks it depended on.  Do we get all that out of the box by using the `depends_on` attribute, or do we have to do anything special to ensure that files persist across tasks and are readable by downstream tasks?

**Comment by fkorotkov** [comment]
Cirrus doesn't support passing working directory around tasks at the moment. There seems to be a lot of small details that are not clear: how to handle re-runs, how to pass the same folder to several tasks in parallel, etc.

The only similar option Cirrus supports now is [Docker Pipe](https://cirrus-ci.org/guide/docker-pipe/) which allows to use several containers within the same tasks but doesn't see like it what you need.

Another option to consider can be using [artifacts](https://cirrus-ci.org/guide/writing-tasks/#artifacts-instruction) and combine 1 & 2 in a separate task that produces an archive which got reused in other tasks.

In the meantime, I've created an [feature request to support in-memory disks](https://github.com/cirruslabs/cirrus-ci-docs/issues/680). Let's see how it will affect the timing and stability.

**Comment by tvolkert** [comment]
@fkorotkov 

> Another option to consider can be using artifacts and combine 1 & 2 in a separate task that produces an archive which got reused in other tasks.

What means of reuse are you picturing == uploading to some server on the first task, then downloading on subsequent tasks?  Or putting the files in a shared on disk?  Does Cirrus support network addressable storage?

**Comment by fkorotkov** [comment]
You can share files between tasks via [Caches](https://cirrus-ci.org/guide/writing-tasks/#cache-instruction), [Artifacts](https://cirrus-ci.org/guide/writing-tasks/#artifacts-instruction) or [built-in HTTP Cache](https://cirrus-ci.org/guide/writing-tasks/#http-cache). With an artifact you'll be able to download it from another task by `$CIRRUS_BUILD_ID`, for example: 

`https://api.cirrus-ci.com/v1/artifact/build/$CIRRUS_BUILD_ID/<TASK NAME>/<ARTIFACT NAME>/<PATH>`

**Comment by tvolkert** [comment]
Are caches _only_ shared by tasks with the same `$CIRRUS_BUILD_ID`?  If so, then it sounds like they'll allow me to accomplish the plan in https://github.com/flutter/flutter/issues/60646#issuecomment-663953550, no?

**Comment by fkorotkov** [comment]
@tvolkert no, caches are global per repository and uses `fingerprint_script` as a key. Sorry for the confusion, just listed all possible options. 

But probably yeah, you can use something like `echo "Docs $CIRRUS_CHANGE_IN_REPO"` as a `fingerprint_script` for the tasks to share a folder after step 1.

**Comment by tvolkert** [comment]
@fkorotkov I think I have this working in https://github.com/flutter/flutter/pull/62281 - can you sanity check my use of Cirrus's config file in that PR?

**Comment by tvolkert** [comment]
@fkorotkov how come in https://cirrus-ci.com/task/4913444795187200, it says that it skipped change detection for docs cache (in the "Upload 'docs' cache" section at the end)?

**Comment by imWildCat** [comment]
@tvolkert  

> Try to find a way to speed up the dashing process. @imWildCat given your help with the original wiring up of this, any chance you have insight here?

Thanks for tagging me!

I haven't looked into the logic into [dashing](https://github.com/technosophos/dashing/blob/master/dashing.go) deeply.

Do you think the timeout issue is due to the increasing size of Flutter's documentation with dart?

If so, I would like to spend my spare time to improve the performance of dashing.

**Comment by fkorotkov** [comment]
@tvolkert since you have the fingerprint script Cirrus assumes that your cache is pure and the same fingerprint will result in the same cache config. You can change the behaviour by setting `reupload_on_changes: true` for the cache instruction, then Cirrus will check for modifications, additions and removals of files inside the caching folder.

**Comment by fkorotkov** [comment]
I've also enabled using a `tmpfs` disk for working directory by setting `use_in_memory_disk` flag:

```yaml
task:
  name: Much I/O
  container:
    image: alpine:latest
    use_in_memory_disk: true
  ...
```

**Comment by jmagman** [comment]
Relying on Cirrus cache contents concerns me, anecdotally the cache has failed to populate often (see https://github.com/cirruslabs/cirrus-ci-docs/issues/660).  Turning off `reupload_on_changes` would be required, but also seems to cause Google Storage read errors when multiple tasks are uploading with the same cache fingerprint.
https://github.com/cirruslabs/cirrus-ci-docs/issues/567.  But maybe all those issues have been resolved.

**Comment by fkorotkov** [comment]
The concurrency thing should be fixed but yeah caches by nature can be missing. I'd suggest to try out [the in-memory disk](https://github.com/cirruslabs/cirrus-ci-docs/pull/683) to have consistent and very fast I/O. As I understood from the Discord discussion it seemed like I/O was very slow which caused the timeout.

**Comment by tvolkert** [comment]
@fkorotkov the in-memory disk didn't seem to help at all -- it anything, the tasks with in-memory disk and more RAM took longer...

### Without in-memory disks:
| Task | Time Taken |
| --- | --- |
| [`docs-linux`](https://cirrus-ci.com/task/5032572524691456) | 40:19 |
[`docs_docset-linux`](https://cirrus-ci.com/task/4680007450624000) | 43:25 |

### _With_ in-memory disks:
| Task | Time Taken |
| --- | --- |
| [`docs-linux`](https://cirrus-ci.com/task/5164581867749376) | 46:33 |
[`docs_docset-linux`](https://cirrus-ci.com/task/6290481774592000) | 58:52 |

**Comment by tvolkert** [comment]
I'm going to disable docset generation while we investigate to unblock the build.

**Comment by tvolkert** [comment]
Note that apropos of the concerns @jmagman brings up about the cache, even though `reupload_on_changes: true` [was set for the docset task](https://github.com/flutter/flutter/blob/39fa00201dfa8340b93e4f2de4fb395510b03b9b/.cirrus.yml#L305), the task execution [still said "Skipping change detection for docs cache!"](https://cirrus-ci.com/task/4680007450624000?command=Upload%20%27docs%27%20cache#L0)

I'm disabling the use of the cache at the same time I disable docset generation.

**Comment by Hixie** [comment]
Supporting docsets isn't P0 so I'm reducing this down to P3.

**Comment by dnfield** [comment]
We could consider using Isoalted from Chrome Infra/LUCI - it's meant for this sort of thing. It allows uploading of arbitrary hashed assets and has some logic to deduplicate/prevent uploads of the same asset over and over again, and is designed for passing around build outputs from one builder to another without introducing unrepeatable dependencies (like a builder-specific disk cache).

**Comment by fkorotkov** [comment]
The [`dashing` library](https://github.com/technosophos/dashing) that is used looks very suspicious. It's single threaded, creates a SQLLite database and inserts very innefiently, recompiles RegEx for every file, etc.

Is it true that only dashing is hanging? If so, it might be related to the inefficient index population.

**Comment by tvolkert** [comment]
Yes, it was the dashing process that was taking way too long.

**Comment by fkorotkov** [comment]
I did a quick pick at the dashing code and added some optimizations [here](https://github.com/fkorotkov/dashing/commit/721088c1e4b583036e796a2f021d6070ea37c487). If you want you can my fork to see if it performs better. Locally id did see a few times faster.

**Comment by christopherfujino** [comment]
> I did a quick pick at the dashing code and added some optimizations [here](https://github.com/fkorotkov/dashing/commit/721088c1e4b583036e796a2f021d6070ea37c487). If you want you can my fork to see if it performs better. Locally id did see a few times faster.

awesome! I made a test draft PR https://github.com/flutter/flutter/pull/62586, if it works we can decide if we want to upstream it and continue to depend on upstream, or just maintain our own fork (or even better, just publish the prebuilt to cipd).

**Comment by zanderso** [comment]
@christopherfujino @fkorotkov If the dashing authors are receptive of pull requests, then I think that would be the first thing to try.

**Comment by fkorotkov** [comment]
Already opened https://github.com/technosophos/dashing/pull/59 but dashing's readme indicates it's in maintenance mode. There is also no license there.

**Comment by christopherfujino** [comment]
> There is also no license there.

Looks like an uber-permissive license: https://github.com/technosophos/dashing/blob/master/LICENSE.txt

**Comment by fkorotkov** [comment]
My bad. For some reason didn't notice it. 🤷‍♂️

**Comment by tvolkert** [comment]
@fkorotkov any progress on upstreaming the fix?

**Comment by fkorotkov** [comment]
Had no response to the PR unfortunately. Also not sure if @christopherfujino had any success with trying out the fork.

**Comment by tvolkert** [comment]
@christopherfujino any luck trying out the fork?

**Comment by Hixie** [comment]
We should make sure we have fixed this before our next Stable, at least. cc @pcsosinski

**Comment by pcsosinski** [comment]
assigning to you @christopherfujino based on Todd's comment above - if you're not the right person please lmk.

**Comment by christopherfujino** [comment]
No luck with the fork. Unassigning myself.

**Comment by tvolkert** [comment]
> No luck with the fork

You mean you were able to run it, and it didn't run any faster, or you weren't able to run it because it failed, or something else?

**Comment by christopherfujino** [comment]
Per this comment https://github.com/flutter/flutter/pull/62586#issuecomment-676712558, it took 54:39 to run (counting setup, 57 minutes). So it doesn't seem any faster.

**Comment by tvolkert** [comment]
@godofredoc how do I see logs for the task?  If I click on "View on Cloud Console" links (e.g. "Task related server Logs"), I get "You do not have permission to view logs."

**Comment by godofredoc** [comment]
You need to click on the Rich Output tab:

![image](https://user-images.githubusercontent.com/54371434/91068059-8da81e00-e5e8-11ea-91a9-f65e952510df.png)

An then click on the execution_details link of on of the steps.

**Comment by tvolkert** [comment]
Thanks!

It looks like the docset generation took ~15m (out of ~37m for the whole task).  👍  for moving this to LUCI and re-enabling the docset generation.

**Comment by pcsosinski** [comment]
what does uploading to firebase validate, what are we actually testing with it?  is there a way we can make this test more hermetic by getting rid of that step or doing it somewhere else?

**Comment by christopherfujino** [comment]
> what does uploading to firebase validate, what are we actually testing with it? is there a way we can make this test more hermetic by getting rid of that step or doing it somewhere else?

Uploading isn't a test, it's publishing the docs to https://master-api.flutter.dev and https://api.flutter.dev.

**Comment by christopherfujino** [comment]
So this shard functions both as an integration test pre-submit, that it is POSSIBLE to generate the docset, and then on post-submit for master and stable channels, it actually publishes the generated docs to the website.

**Comment by pcsosinski** [comment]
IMO it's very confusing to have a "test" like this that tests something on pre submit and then switches to deploying something on post submit - basically an opportunity to pass pre submit and fail the same test on post submit that requires this tribal knowledge to understand why that is happening.

I really think we should consider splitting these two actions into separate (and clearly named ) runs.  it's also important to do for your atomic release plan, @christopherfujino

**Comment by christopherfujino** [comment]
> IMO it's very confusing to have a "test" like this that tests something on pre submit and then switches to deploying something on post submit - basically an opportunity to pass pre submit and fail the same test on post submit that requires this tribal knowledge to understand why that is happening.
> 
> I really think we should consider splitting these two actions into separate (and clearly named ) runs. it's also important to do for your atomic release plan, @christopherfujino

Makes sense to me. @gspencergoog do we still need to run this on pre-submit? Does it actually fail legitimately on pre-submit (I think maybe I remember once it failed because we needed to update dartdoc)?

**Comment by gspencergoog** [comment]
@pcsosinski @christopherfujino Yes, it actually fails on presubmit sometimes, and it definitely needs to run there. It is a test that documentation is correct and can be generated after changes made in source code. The dartdoc build fails if there are any dartdoc warnings which indicate broken links in new documentation. It also fails if there is a bug in the snippet tool, or if the dartdoc version is incompatible with the documentation being parsed from the source code (e.g. if it uses features not present in the version of dartdoc we are using).

If we split the actions into two parts, they will be virtually identical.  The only thing that the post-submit one does that the test wouldn't is actually publish the docs, which, for obvious reasons, we don't want the pre-submit to do. To do that, we would implement it by calling the same script with different environment variables or command line options, which is exactly what we do today.

If you look in `docs.sh`, the entire script needs to run for the test except for the last stanza that does the publish if the appropriate conditions are met (i.e. that it is running post-submit).

We can simply rename the step to not be called a "test" when it runs post-submit, but that just complicates the build rules, because we'd have to exclude the test step on post-submit, and exclude the publish step on pre-submit, and run the same command in both cases, which would be confusing when reading the build config ("why all this complicated logic to run the same command either way?").

**Comment by gspencergoog** [comment]
Perhaps we can come up with a single name that describes the step more clearly?  Maybe `try_publish_docs`?

**Comment by tvolkert** [comment]
@godofredoc what's the current status of this?

**Comment by godofredoc** [comment]
I still need to send the PRs for review. I'll send then before the end of the week.

**Comment by godofredoc** [comment]
Test has been migrated to LUCI and docset is being generated. However seems like the script is not generating the offline docs and I'm not sure how/where to validate the dashing docs.

@gspencergoog @xster from comments seems like you are the ones using/maintaining those doc. Can you please take a look and what is missing?

**Comment by gspencergoog** [comment]
The dashing docs can be validated in [Zeal(Linux/Windows)](https://zealdocs.org/) or [Dash(macOS)](https://kapeli.com/dash). Where do I find the logs from the generation run so I can see what it's doing?

**Comment by godofredoc** [comment]
This is the task:

https://ci.chromium.org/p/flutter/builders/prod/Linux%20docs/627

This is the step generating the docs:

https://logs.chromium.org/logs/flutter/buildbucket/cr-buildbucket.appspot.com/8866433229007858752/+/steps/prepare_environment/0/steps/Docs/0/steps/docs/0/stdout

And this is the step uploading to firebase:

https://logs.chromium.org/logs/flutter/buildbucket/cr-buildbucket.appspot.com/8866433229007858752/+/steps/prepare_environment/0/steps/Docs/0/steps/Deploy_docs/0/steps/Firebase_deploy/0/stdout

**Comment by gspencergoog** [comment]
@godofredoc It looks like it actually is generating the Dash/Zeal docset properly, as well as the offline zip file version of the docs. The problem only seems to be the 401's (Unauthorized) coming from the Firebase server, which would imply that the auth token isn't properly set up for Firebase.

**Comment by gspencergoog** [comment]
Here's where I can see that it generated the docset: it's in [this log](https://logs.chromium.org/logs/flutter/buildbucket/cr-buildbucket.appspot.com/8866433229007858752/+/steps/prepare_environment/0/steps/Docs/0/steps/docs/0/stdout), at the very bottom.

**Comment by godofredoc** [comment]
Perfect seems like is just the log lines:
```
430M	doc/offline/flutter.docs.zip
313M	doc/offline/flutter.docset.tar.gz
Wed Oct 14 21:08:50 UTC 2020: Updating master docs: https://master-api.flutter.dev/
```
Made me think the packages were on https://master-api.flutter.dev/doc/offline/flutter.docs.zip but they are actually 
in https://master-api.flutter.dev/offline/flutter.docset.tar.gz

**Comment by godofredoc** [comment]
I think we are good to go, thanks @gspencergoog for checking. I'll close this bug and reopen if it is needed.

**Comment by godofredoc** [comment]
By the way I checked in the firebase console and we have several successful uploads.

**Comment by gspencergoog** [comment]
Yeah, it would have printed errors if it failed.  Those lines are printed to verify that it indeed produced output (and to track the size).
